Use node-pre-gyp to publish binaries

ksmyth opened the pull request:
This PR adds node-pre-gyp support.

Compiling modules is a pain on Windows, since every machine needs the compiler. Downloading pre-built binaries during `npm install` is a big convenience.

I made a v0.8.2 release to test:

    npm install bcrypt@https://github.com/ksmyth/node.bcrypt.js/tarball/v0.8.2
    > bcrypt@0.8.2 install C:\Users\kevin\Documents\webgme\node_modules\bcrypt
    > node-pre-gyp install --fallback-to-build

    [bcrypt] Success: "C:\Users\kevin\Documents\webgme\node_modules\bcrypt\lib\binding\bcrypt_lib.node" is installed via remote
    bcrypt@0.8.2 node_modules\bcrypt
    ├── bindings@1.0.0
    ├── nan@1.5.0
    └── node-pre-gyp@0.6.4

The log indicates that `https://github.com/ksmyth/node.bcrypt.js/releases/download/v0.8.2/bcrypt_lib-v0.8.2-node-v11-win32-x64.tar.gz` was downloaded and installed.

If you are open to distributing binaries, publishing the binaries could be automated with e.g. appveyor.

Globegitter commented:
:+1: It is also great to have pre-published binaries when using minimal docker files for iojs or node.

mrgamer commented:
:+1: plus one to this, is troublesome to use a different bcrypt library only to develop on windows

Wargog commented:
:+1: Please pull

ksmyth commented:
I've updated ksmyth/master to the latest release.

    npm install bcrypt@https://github.com/ksmyth/node.bcrypt.js/tarball/v0.8.5

I've also added an appveyor config. The binaries have been downloaded from appveyor and uploaded to github.

vjpr commented:
+1

ksmyth commented:
I added Node 5 (aka node-v47) binaries

jhiswin commented:
Hi, I'm trying to use this but it always seems to try to fall back to using node-gyp to compile.  Am I doing something wrong?

ksmyth commented:
@jhiswin What version OS, node, and npm are you using? I have only published Windows binaries.

Works for me on Windows 10:

    C:\Users\kevin\tmp\tt>npm "install" "bcrypt@https://github.com/ksmyth/node.bcrypt.js/tarball/v0.8.5"

    > bcrypt@0.8.5 install C:\Users\kevin\tmp\tt\node_modules\bcrypt
    > node-pre-gyp install --fallback-to-build

    [bcrypt] Success: "C:\Users\kevin\tmp\tt\node_modules\bcrypt\lib\binding\bcrypt_lib.node" is installed via remotett@1.0.0 C:\Users\kevin\tmp\tt
    `-- bcrypt@0.8.5  extraneous

    npm WARN tt@1.0.0 No description
    npm WARN tt@1.0.0 No repository field.

    C:\Users\kevin\tmp\tt>npm --version
    3.7.3

    C:\Users\kevin\tmp\tt>node --version
    v5.9.1

defunctzombie commented:
If we wanted to ship binaries, what service should be used to build them? I don't have windows and would want the binaries automatically built.

ksmyth commented:
Appveyor is building binaries for my fork:
https://ci.appveyor.com/project/ksmyth/node-bcrypt-js

It does not automatically upload builds for tags to GitHub, but appveyor has docs on how to do this.

defunctzombie commented:
@ksmyth do these work across various windows versions? Would you be willing to update this PR so we can take another stab at it? I'll be more responsive :)

ksmyth commented:
@defunctzombie I've updated this PR with a squashed commit.

Note it includes a github.com/ksmyth URL for the published binaries. I'd be willing to upload them, or you can change it and pull binaries from appveyor when you make a release.

ksmyth commented:
`node-pre-gyp` requires Node >= 0.10. Are you willing to drop support for the very old Node versions?

defunctzombie commented:
Yes, I think dropping < 0.10 is going to be ok.

defunctzombie commented on package.json:
```diff
@@ -25,12 +25,17 @@
+    "node-pre-gyp": "0.6.4"
   },
+  "bundledDependencies": [
+    "node-pre-gyp"
```
does this need to be bundle dep?

ksmyth replied:
Hmm, looks like it changed here https://github.com/mapbox/node-pre-gyp/commit/8f7a774980a6aee1f4d4d2c36d8a27e2d3841334#diff-04c6e90faac2675aa89e2176d2eec7d8

I will remove it and test.

ksmyth commented:
I pushed and tested a new commit. I removed the `devDependency` and bumped the `node-pre-gyp` version.

defunctzombie commented on package.json:
```diff
@@ -49,5 +51,11 @@
+    "module_name": "bcrypt_lib",
+    "module_path": "./lib/binding/",
+    "host": "https://github.com",
+    "remote_path": "/ksmyth/node.bcrypt.js/releases/download/v{version}/"
```
do you know if there is a way to reference appveyor directly? Seems a bit of a chicken/egg problem. If I want to make binaries for some version `v`, appveyor will only generate the binary for that after I make the commit bumping the version but then I will need to update the package.json file with the new url to that version?

I suppose one option is to push to github, wait for the appveyor build, download the binary and upload to github releases, and then publish to npm. Is this the workflow you currently use?

ksmyth replied:
`remote_path` substitutes in the current version number, so you don't need to change the URL.

That sounds like a reasonable workflow.

I haven't published this to npm. The workflow I use is to `npm install [full URL to the release .tgz]` after uploading the binaries. Then `npm install` to install everything else.

defunctzombie replied:
Great, I have created the appveyor account. Could you update this url to reference `ncb000gt` vs `ksmyth` and I will merge. I could do this with some git gymnastics but figured would be easiest if done here and PR merged per normal methods.

ksmyth commented:
I changed the URL to `ncb00gt` and pushed.

defunctzombie commented:
@ksmyth thank you! Do you know if travis will continue to rebuild for PRs or will try to use the pre-built binaries? Presumably this won't break linux or windows users if those binaries are not available?

Sequoia commented:
How do I find a list of what binaries are available and what platforms/versions they support? I'd like to simplify my deploy (and dev builds on mac/windows/linux) by targeting a bcrypt+node version with pre-built binaries.

ksmyth commented:
@Sequoia the binaries are listed at https://github.com/kelektiv/node.bcrypt.js/releases

v1.0.2 is the only version with binaries, and only for Windows.

Maybe @defunctzombie would be open to a PR that added `node-pre-gyp package` to `.travis.yml`

Sequoia commented:
Thanks! Is there a way for me to find out what engines/platforms each of those support?
